apigee/registry
# Implement a sample linter for OpenAPI that enforces rules on descriptions

## Summary
The following pull request creates a sample OpenAPI linter which serves as a proof-of-concept for custom linter creation in the API Registry. It lints two rules:
- `description-less-than-1000-chars`: Ensures that description fields in the OpenAPI spec are less than 1000 characters.
- `description-contains-no-tags`: Ensures that description fields in the OpenAPI spec contain no tags (such as HTML tags).

## Test
- Have registry running on `localhost:8080`.
- Place [the following OpenAPI Sample spec](https://gist.github.com/muhammadharis/13f1681ad65bf6a59366f3068748c0d8) into ~/petstore.yaml 
- Run `LOCAL.sh`
- Ensure that the following style guide is located under `~/styleguide.yaml`:
```yaml
id: openapistyleguide
mime_types:
  - application/x.openapi+gzip;version=2
  - application/x.openapi+gzip;version=3
guidelines:
  - id: Description
    display_name: Govern Description Properties
    description: This guideline governs properties for ref fields on specs.
    rules:
      - id: DescriptionLessThan1000Chars
        description: Description fields should be less than 1000 chars
        linter: openapi-sample
        linter_rulename: description-less-than-1000-chars
        severity: ERROR

      - id: DescriptionContainsNoTags
        description: Description fields should not contain tags
        linter: openapi-sample
        linter_rulename: description-contains-no-tags
        severity: ERROR
    status: ACTIVE
linters:
  - name: openapi-sample
```


Run the following series of commands:

```bash
apg admin delete-project --name projects/example

apg admin create-project --project_id example --insecure --address localhost:8080

apg registry create-api --api_id petstore --parent=projects/example/locations/global

apg registry create-api-version --api_version_id v1 --parent=projects/example/locations/global/apis/petstore

registry upload spec --version=projects/example/locations/global/apis/petstore/versions/v1 --style=openapi ~/petstore.yaml

registry upload styleguide ~/styleguide.yaml --project-id=example

registry compute conformance projects/test/locations/global/apis/petstore/versions/v1/specs/petstore.yaml

registry get projects/example/locations/global/apis/petstore/versions/v1/specs/petstore.yaml/artifacts/conformance-openapistyleguide --contents
```

Ensure that the output matches [the following gist](https://gist.github.com/muhammadharis/7ed03bf240d6b34cbcf97a89f6c2718e).

## Running Unit Tests
`go test -timeout 30s github.com/apigee/registry/cmd/registry/plugins/registry-lint-openapi-sample`
